app/common/expressions/__init__.py

```diff
@@ -0,0 +1,101 @@
+import ast
+import uuid
+from typing import Any
+
+import simpleeval
+
+from app.common.data.models import Expression
+
+
+class BaseExpressionError(Exception):
+    pass
+
+
+class UndefinedVariableInExpression(BaseExpressionError):
+    pass
+
+
+class DisallowedExpression(BaseExpressionError):
+    pass
+
+
+class InvalidEvaluationResult(BaseExpressionError):
+    pass
+
+
+def _evaluate_expression_with_context(
+    expression: Expression, context: dict[str, str | int | float | bool | None] | None = None
+) -> Any:
+    """
+    The base evaluator to use for handling all expressions.
+
+    This parses arbitrary Python-language text into an Abstract Syntax Tree (AST) and then evaluates the result of
+    that expression. Parsing arbitrary Python is extremely dangerous so we heavily restrict the AST nodes that we
+    are willing to handle, to (hopefully) close off the attack surface to any malicious behaviour.
+
+    The addition of any new AST nodes should be well-tested and intentional consideration should be given to any
+    ways of exploit or misuse.
+    """
+    expr_context = expression.context or {}
+    context = context or {}
+
+    if context_overlap := set(expr_context).intersection(set(context)):
```

Comment:
This is a really great guard and feels like it will have a great developer experience in the failure if we come across it 👍 great forward thinking